<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.553">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Seurat_spatialvignette</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="Seurat_Visium_tutorial_files/libs/clipboard/clipboard.min.js"></script>
<script src="Seurat_Visium_tutorial_files/libs/quarto-html/quarto.js"></script>
<script src="Seurat_Visium_tutorial_files/libs/quarto-html/popper.min.js"></script>
<script src="Seurat_Visium_tutorial_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="Seurat_Visium_tutorial_files/libs/quarto-html/anchor.min.js"></script>
<link href="Seurat_Visium_tutorial_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="Seurat_Visium_tutorial_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="Seurat_Visium_tutorial_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="Seurat_Visium_tutorial_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="Seurat_Visium_tutorial_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">


</head>

<body class="fullcontent">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">

<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Seurat_spatialvignette</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<section id="introduction-to-visium-data-in-r-using-seurat" class="level2">
<h2 class="anchored" data-anchor-id="introduction-to-visium-data-in-r-using-seurat">Introduction to Visium Data in R using Seurat</h2>
<p>BCBB Workshop for NIH Researchers, Nov 2024</p>
<p>We will be using the packages <strong>Seurat</strong> and <strong>ggplot2</strong> for visualization. Example data comes from <strong>SeuratData</strong> package. Tutorial follows <a href="https://satijalab.org/seurat/articles/spatial_vignette">Seurat official vignette</a> but adds additional code, demo of count data and explanations.</p>
<hr>
<section id="setup-ideally-run-before-class-to-install-packages-load-libraries-and-download-data-ahead-of-time" class="level3">
<h3 class="anchored" data-anchor-id="setup-ideally-run-before-class-to-install-packages-load-libraries-and-download-data-ahead-of-time">SETUP: Ideally, Run Before Class (to install packages, load libraries, and download data ahead of time)</h3>
<p>In <span style="color:red">red</span> is setup that you can run prior to the tutorial to make sure you have the R packages and data ahead of time to make sure the tutorial runs most smoothly.</p>
<p>Note that some commands may have a comment mark <span style="color:red">#</span> that you will need to remove in order to run them.</p>
<p><span style="color:red">Install the packages:</span></p>
<p><span style="color:red"><strong>remotes</strong> and <strong>devtools</strong> are useful tools for installing packages</span></p>
<p><span style="color:red"><strong>Seurat</strong> (required) and <strong>Seurat Data</strong> (only necessary for the tutorial)</span></p>
<p><span style="color:red">You will also need <strong>ggplot</strong> for plotting and <strong>dplyr</strong>, <strong>tidyr</strong> for data analysis</span></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co">#install.packages('remotes')</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="do">##remotes::install_github("satijalab/seurat", "seurat5", quiet = TRUE)</span></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="do">##remotes::install_github("satijalab/seurat-data", "seurat5", quiet = TRUE)</span></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="do">##remotes::install_github("satijalab/azimuth", "seurat5", quiet = TRUE)</span></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="do">##remotes::install_github("satijalab/seurat-wrappers", "seurat5", quiet = TRUE)</span></span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="do">##remotes::install_github("stuart-lab/signac", "seurat5", quiet = TRUE)</span></span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="do">##remotes::install_github("bnprks/BPCells", quiet = TRUE)</span></span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="co">#install.packages('Seurat')</span></span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="co">#install.packages('devtools')</span></span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a><span class="co">#devtools::install_github('satijalab/seurat-data')</span></span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a><span class="co">#install.packages("ggplot2")</span></span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a><span class="co">#install.packages("patchwork")</span></span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a><span class="co">#install.packages("dplyr")</span></span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a><span class="co">#install.packages("tidyr")</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><span style="color:red">We will need the following libraries:</span></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(SeuratData)</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(patchwork)</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyr)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><span style="color:red"><strong>Download the mouse brain Visium dataset ahead of time</strong></span></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co">#InstallData("stxBrain")</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><span style="color:red">Later, for running RCTD, you will also need the spacexr R package and a mouse cortex reference</span>. Note that I had issues downloading until I took myself off NIH VPN (try that if it doesn’t work for you on VPN).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co">#if (!requireNamespace("spacexr", quietly = TRUE)) {</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="co">#  devtools::install_github("dmcable/spacexr", build_vignettes = FALSE)</span></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a><span class="co">#}</span></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a><span class="do">#### OR </span></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a><span class="co">#remotes::install_github("dmcable/RCTD")</span></span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(spacexr)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><span style="color:red"><strong>Download mouse cortex reference <a href="https://www.dropbox.com/scl/fi/r1mixf4eof2cot891n215/allen_scRNAseq_ref.Rds?rlkey=ynr6s6wu1efqsjsu3h40vitt7&amp;dl=0">here</a>.</strong> This reference scRNAseq dataset has been reduced to 200,000 cells (and rare cell types fewer than 25 cells have been removed).</span></p>
<p><span style="color:red">We will be running a bunch of calculations which will be sped up with the <strong>glmGampoi</strong> and <strong>Rfast2</strong> packages, so let’s install and load them as well:</span></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co">#BiocManager::install('glmGamPoi')</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="do">### fits Gamma-Poisson (aka Negative Binomial) Generalized Linear Models </span></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a><span class="do">### substantially improves speed for the SCTransform algorithm</span></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(glmGamPoi)</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a><span class="co">#install.packages("Rfast2")</span></span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a><span class="do">### substantially improves speed for  FindSpatialVariableFeatures</span></span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Rfast2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now you should now be ready to run the rest of the tutorial fairly quickly!</p>
<hr>
</section>
<section id="useful-resources-for-seurat" class="level3">
<h3 class="anchored" data-anchor-id="useful-resources-for-seurat">Useful Resources for Seurat</h3>
<p>Useful resources for understanding the Seurat object class and methods:</p>
<p>Seurat wiki: <a href="https://github.com/satijalab/seurat/wiki" class="uri">https://github.com/satijalab/seurat/wiki</a></p>
<p>Essential Seurat commands: <a href="https://satijalab.org/seurat/articles/essential_commands.html" class="uri">https://satijalab.org/seurat/articles/essential_commands.html</a></p>
<p>Every Seurat object consists as a collection of Assay objects and DimReduc objects.</p>
<p>Each Seurat object has slots:</p>
<table class="table">
<colgroup>
<col style="width: 36%">
<col style="width: 63%">
</colgroup>
<thead>
<tr class="header">
<th>Slot</th>
<th>Function</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><code>assays</code></td>
<td>A list of assays within this object</td>
</tr>
<tr class="even">
<td><code>meta.data</code></td>
<td>Cell-level meta data</td>
</tr>
<tr class="odd">
<td><code>active.assay</code></td>
<td>Name of active, or default, assay</td>
</tr>
<tr class="even">
<td><code>active.ident</code></td>
<td>Identity classes for the current object</td>
</tr>
<tr class="odd">
<td><code>graphs</code></td>
<td>A list of nearest neighbor graphs</td>
</tr>
<tr class="even">
<td><code>reductions</code></td>
<td>A list of DimReduc objects</td>
</tr>
<tr class="odd">
<td><code>project.name</code></td>
<td>User-defined project name (optional)</td>
</tr>
<tr class="even">
<td><code>tools</code></td>
<td>Empty list. Tool developers can store any internal data from their methods here</td>
</tr>
<tr class="odd">
<td><code>misc</code></td>
<td>Empty slot. User can store additional information here</td>
</tr>
<tr class="even">
<td><code>version</code></td>
<td>Seurat version used when creating the object</td>
</tr>
</tbody>
</table>
<p>You can also check the list of methods available for Seurat objects like so:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>utils<span class="sc">::</span><span class="fu">methods</span>(<span class="at">class =</span> <span class="st">'Seurat'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] [                             [[                           
 [3] [[&lt;-                          $                            
 [5] $&lt;-                           AddMetaData                  
 [7] as.CellDataSet                as.SingleCellExperiment      
 [9] Assays                        CastAssay                    
[11] Cells                         colMeans                     
[13] colSums                       Command                      
[15] DefaultAssay                  DefaultAssay&lt;-               
[17] DefaultFOV                    DefaultFOV&lt;-                 
[19] dim                           dimnames                     
[21] dimnames&lt;-                    droplevels                   
[23] Embeddings                    Features                     
[25] FetchData                     FindClusters                 
[27] FindMarkers                   FindNeighbors                
[29] FindSpatiallyVariableFeatures FindVariableFeatures         
[31] FoldChange                    GetAssay                     
[33] GetAssayData                  GetImage                     
[35] GetTissueCoordinates          head                         
[37] HVFInfo                       Idents                       
[39] Idents&lt;-                      initialize                   
[41] JoinLayers                    Key                          
[43] Keys                          LayerData                    
[45] LayerData&lt;-                   Layers                       
[47] levels                        levels&lt;-                     
[49] LeverageScore                 Loadings                     
[51] merge                         Misc                         
[53] Misc&lt;-                        names                        
[55] NormalizeData                 Project                      
[57] Project&lt;-                     ProjectCellEmbeddings        
[59] ProjectUMAP                   PseudobulkExpression         
[61] RenameCells                   RenameIdents                 
[63] ReorderIdent                  rowMeans                     
[65] rowSums                       RunCCA                       
[67] RunGraphLaplacian             RunICA                       
[69] RunLDA                        RunPCA                       
[71] RunSLSI                       RunSPCA                      
[73] RunTSNE                       RunUMAP                      
[75] ScaleData                     ScoreJackStraw               
[77] SCTransform                   SCTResults                   
[79] SetAssayData                  SetIdent                     
[81] show                          SpatiallyVariableFeatures    
[83] split                         StashIdent                   
[85] Stdev                         subset                       
[87] SVFInfo                       tail                         
[89] Tool                          Tool&lt;-                       
[91] VariableFeatures              VariableFeatures&lt;-           
[93] Version                       WhichCells                   
see '?methods' for accessing help and source code</code></pre>
</div>
</div>
<hr>
</section>
<section id="load-visium-data" class="level3">
<h3 class="anchored" data-anchor-id="load-visium-data">Load Visium Data</h3>
<p>Let’s load some stxBrain Visium data from the SeuratData package</p>
<p>This data consists of a single anterior slice named <code>anterior1</code></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>?stxBrain</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="co">#InstallData("stxBrain")</span></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>brain <span class="ot">&lt;-</span> <span class="fu">LoadData</span>(<span class="st">"stxBrain"</span>, <span class="at">type =</span> <span class="st">"anterior1"</span>)</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>brain<span class="sc">@</span>assays</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$Spatial
Assay (v5) data with 31053 features for 2696 cells
First 10 features:
 Xkr4, Gm1992, Gm37381, Rp1, Sox17, Gm37323, Mrpl15, Lypla1, Gm37988,
Tcea1 
Layers:
 counts </code></pre>
</div>
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Cells(brain)</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="co">#Features(brain)</span></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a><span class="co">#brain$nCount_Spatial </span><span class="al">###</span><span class="co"> numbers of UMIs per spot</span></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a><span class="co">#brain$nFeature_Spatial </span><span class="al">###</span><span class="co"> number of features detected per spot</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We are using a convenient package here that includes pre-loaded data, but for a typical Visium experiment we would use the following command</p>
<pre><code>Load10x_Spatial(dir=“/data/dir/”, filename = "filtered_feature_bc_matrix.h5", assay = "Spatial", slice = "slice1")</code></pre>
<p>For additional help to load custom 10X Visium data into a Seurat object, see <a href="https://satijalab.org/seurat/reference/load10x_spatial" class="uri">https://satijalab.org/seurat/reference/load10x_spatial</a></p>
<p>This 10X Visium data consists of:</p>
<ul>
<li><p>A spot by gene expression matrix</p></li>
<li><p>An image of the tissue slice (obtained from H&amp;E staining during data acquisition)</p></li>
<li><p>Scaling factors that relate the original high resolution image to the lower resolution image used here for visualization.</p></li>
</ul>
<p>In this Seurat object containing 10X Visium STx data the spot by gene expression matrix contains spot level, not single-cell level data. The “images” slot of the Seurat object contains the image of the tissue as well spatial information needed to associate spot with spatial coordinates in the tissue image.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">nrow</span>(brain) <span class="co"># the number of features / genes</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 31053</code></pre>
</div>
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ncol</span>(brain) <span class="co"># the number of samples / cells / spots</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 2696</code></pre>
</div>
</div>
<p><strong>glimpse()</strong> is as a handy way to quickly look at an object in R (gives the hierarchy/data structure):</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="fu">glimpse</span>(brain)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Formal class 'Seurat' [package "SeuratObject"] with 13 slots
  ..@ assays      :List of 1
  .. ..$ Spatial:Formal class 'Assay5' [package "SeuratObject"] with 8 slots
  ..@ meta.data   :'data.frame':    2696 obs. of  5 variables:
  .. ..$ orig.ident      : Factor w/ 1 level "anterior1": 1 1 1 1 1 1 1 1 1 1 ...
  .. ..$ nCount_Spatial  : num [1:2696] 13069 37448 28475 39718 33392 ...
  .. ..$ nFeature_Spatial: int [1:2696] 4242 7860 6332 7957 7791 6291 7487 7043 4876 8985 ...
  .. ..$ slice           : num [1:2696] 1 1 1 1 1 1 1 1 1 1 ...
  .. ..$ region          : chr [1:2696] "anterior" "anterior" "anterior" "anterior" ...
  ..@ active.assay: chr "Spatial"
  ..@ active.ident: Factor w/ 1 level "anterior1": 1 1 1 1 1 1 1 1 1 1 ...
  .. ..- attr(*, "names")= chr [1:2696] "AAACAAGTATCTCCCA-1" "AAACACCAATAACTGC-1" "AAACAGAGCGACTCCT-1" "AAACAGCTTTCAGAAG-1" ...
  ..@ graphs      : list()
  ..@ neighbors   : list()
  ..@ reductions  : list()
  ..@ images      :List of 1
  .. ..$ anterior1:Formal class 'VisiumV2' [package "Seurat"] with 6 slots
  ..@ project.name: chr "anterior1"
  ..@ misc        : list()
  ..@ version     :Classes 'package_version', 'numeric_version'  hidden list of 1
  .. ..$ : int [1:3] 5 0 2
  ..@ commands    : list()
  ..@ tools       : list()</code></pre>
</div>
</div>
<hr>
</section>
<section id="seurat-object-metadata" class="level3">
<h3 class="anchored" data-anchor-id="seurat-object-metadata">Seurat object metadata</h3>
<p>Metadata is stored as a <code>data.frame</code>, where each row is a sample (e.g.&nbsp;cell or spot) and each column correspond to one sample-level metadata field.</p>
<p>Accessed via <code>[[</code> extract operator, the <code>meta.data</code> object, or the <code>$</code> sigil (<code>$</code> extracts one single column at a time). Row names in the metadata need to match the column names of the counts matrix.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(brain<span class="sc">@</span>meta.data)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                   orig.ident nCount_Spatial nFeature_Spatial slice   region
AAACAAGTATCTCCCA-1  anterior1          13069             4242     1 anterior
AAACACCAATAACTGC-1  anterior1          37448             7860     1 anterior
AAACAGAGCGACTCCT-1  anterior1          28475             6332     1 anterior
AAACAGCTTTCAGAAG-1  anterior1          39718             7957     1 anterior
AAACAGGGTCTATATT-1  anterior1          33392             7791     1 anterior
AAACATGGTGAGAGGA-1  anterior1          20955             6291     1 anterior</code></pre>
</div>
</div>
<p>Here we can look at the distribution of transcripts detected per spot and spatially across the sample image:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>brainCount_Spatial_VlnPlot <span class="ot">&lt;-</span> <span class="fu">VlnPlot</span>(brain, <span class="at">features =</span> <span class="st">"nCount_Spatial"</span>, <span class="at">pt.size =</span> <span class="fl">0.1</span>) <span class="sc">+</span> <span class="fu">NoLegend</span>()</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>brain_SpatialFeaturePlot <span class="ot">&lt;-</span> <span class="fu">SpatialFeaturePlot</span>(brain, <span class="at">features =</span> <span class="st">"nCount_Spatial"</span>, <span class="at">images=</span><span class="st">"anterior1"</span>) <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"right"</span>)</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a><span class="fu">wrap_plots</span>(brainCount_Spatial_VlnPlot, brain_SpatialFeaturePlot)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Seurat_Visium_tutorial_files/figure-html/unnamed-chunk-9-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<hr>
</section>
<section id="exploring-what-we-mean-by-overdispersion-of-count-data" class="level3">
<h3 class="anchored" data-anchor-id="exploring-what-we-mean-by-overdispersion-of-count-data">Exploring what we mean by Overdispersion of Count Data</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>nCount_Spatial <span class="ot">&lt;-</span> brain<span class="sc">@</span>meta.data<span class="sc">$</span>nCount_Spatial</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a data frame for plotting</span></span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a>nCount_Spatialdf <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">nCount_Spatial =</span> nCount_Spatial)</span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot histogram to visualize overdispersion</span></span>
<span id="cb21-7"><a href="#cb21-7" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(nCount_Spatialdf, <span class="fu">aes</span>(<span class="at">x =</span> nCount_Spatial)) <span class="sc">+</span></span>
<span id="cb21-8"><a href="#cb21-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">binwidth =</span> <span class="dv">100</span>, <span class="at">fill =</span> <span class="st">"blue"</span>, <span class="at">color =</span> <span class="st">"black"</span>, <span class="at">alpha =</span> <span class="fl">0.7</span>) <span class="sc">+</span></span>
<span id="cb21-9"><a href="#cb21-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Histogram of nCount_Spatial"</span>,</span>
<span id="cb21-10"><a href="#cb21-10" aria-hidden="true" tabindex="-1"></a>       <span class="at">x =</span> <span class="st">"nCount_Spatial"</span>,</span>
<span id="cb21-11"><a href="#cb21-11" aria-hidden="true" tabindex="-1"></a>       <span class="at">y =</span> <span class="st">"Frequency"</span>) <span class="sc">+</span></span>
<span id="cb21-12"><a href="#cb21-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Seurat_Visium_tutorial_files/figure-html/unnamed-chunk-10-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<section id="compare-to-poisson-distribution-of-the-same-mean" class="level4">
<h4 class="anchored" data-anchor-id="compare-to-poisson-distribution-of-the-same-mean">Compare to Poisson Distribution of the same mean</h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="co">#install.packages("MASS")</span></span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(MASS)</span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a><span class="co">#class(nCount_Spatial)</span></span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a>lambda <span class="ot">&lt;-</span> <span class="fu">mean</span>(nCount_Spatialdf<span class="sc">$</span>nCount_Spatial)</span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a>nCount_Spatial_rPoisson<span class="ot">&lt;-</span><span class="fu">rpois</span>(<span class="at">n =</span> <span class="fu">length</span>(nCount_Spatialdf<span class="sc">$</span>nCount_Spatial), lambda)</span>
<span id="cb22-6"><a href="#cb22-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-7"><a href="#cb22-7" aria-hidden="true" tabindex="-1"></a>nCount_df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">nCount_Spatial1 =</span> nCount_Spatial, <span class="at">nCount_SpatialPoisson =</span> nCount_Spatial_rPoisson)</span>
<span id="cb22-8"><a href="#cb22-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-9"><a href="#cb22-9" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb22-10"><a href="#cb22-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">data =</span> nCount_df, <span class="fu">aes</span>(<span class="at">x =</span> nCount_Spatial, <span class="at">fill =</span> <span class="st">"nCount_Spatial"</span>), <span class="at">binwidth =</span> <span class="dv">200</span>, <span class="at">alpha =</span> <span class="fl">0.5</span>, <span class="at">position =</span> <span class="st">'identity'</span>) <span class="sc">+</span></span>
<span id="cb22-11"><a href="#cb22-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">data =</span> nCount_df, <span class="fu">aes</span>(<span class="at">x =</span> nCount_SpatialPoisson, <span class="at">fill =</span> <span class="st">"nCount_SpatialPoisson"</span>), <span class="at">binwidth =</span> <span class="dv">200</span>, <span class="at">alpha =</span> <span class="fl">0.5</span>, <span class="at">position =</span> <span class="st">'identity'</span>) <span class="sc">+</span></span>
<span id="cb22-12"><a href="#cb22-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_manual</span>(<span class="at">name =</span> <span class="st">"Legend"</span>, <span class="at">values =</span> <span class="fu">c</span>(<span class="st">"nCount_Spatial"</span> <span class="ot">=</span> <span class="st">"blue"</span>, <span class="st">"nCount_SpatialPoisson"</span> <span class="ot">=</span> <span class="st">"red"</span>)) <span class="sc">+</span></span>
<span id="cb22-13"><a href="#cb22-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Overlayed Histogram of nCount_Spatial and nCount_SpatialPoisson"</span>, <span class="at">x =</span> <span class="st">"Value"</span>, <span class="at">y =</span> <span class="st">"Count"</span>) <span class="sc">+</span> <span class="fu">theme_minimal</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Seurat_Visium_tutorial_files/figure-html/unnamed-chunk-11-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>The Poisson distribution (equivalent to normal distribution at high n) is really not appropriate for modeling count data! The observed count data has much higher variance.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">data =</span> nCount_df, <span class="fu">aes</span>(<span class="at">x =</span> nCount_SpatialPoisson, <span class="at">fill =</span> <span class="st">"nCount_SpatialPoisson"</span>), <span class="at">binwidth =</span> <span class="dv">10</span>, <span class="at">alpha =</span> <span class="fl">0.5</span>, <span class="at">position =</span> <span class="st">'identity'</span>) <span class="sc">+</span></span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_manual</span>(<span class="at">name =</span> <span class="st">"Legend"</span>, <span class="at">values =</span> <span class="fu">c</span>(<span class="st">"nCount_SpatialPoisson"</span> <span class="ot">=</span> <span class="st">"red"</span>)) <span class="sc">+</span> <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Histogram of nCount_SpatialPoisson"</span>, <span class="at">x =</span> <span class="st">"Value"</span>, <span class="at">y =</span> <span class="st">"Count"</span>) <span class="sc">+</span> <span class="fu">theme_minimal</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Seurat_Visium_tutorial_files/figure-html/unnamed-chunk-12-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">data =</span> nCount_df, <span class="fu">aes</span>(<span class="at">x =</span> nCount_Spatial, <span class="at">fill =</span> <span class="st">"nCount_Spatial"</span>), </span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a>                 <span class="at">binwidth =</span> <span class="dv">200</span>, <span class="at">alpha =</span> <span class="fl">0.5</span>, <span class="at">position =</span> <span class="st">'identity'</span>) <span class="sc">+</span></span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_manual</span>(<span class="at">name =</span> <span class="st">"Legend"</span>, <span class="at">values =</span> <span class="fu">c</span>(<span class="st">"nCount_Spatial"</span> <span class="ot">=</span> <span class="st">"blue"</span>)) <span class="sc">+</span></span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Histogram of nCount_Spatial"</span>,</span>
<span id="cb24-6"><a href="#cb24-6" aria-hidden="true" tabindex="-1"></a>       <span class="at">x =</span> <span class="st">"Value"</span>,</span>
<span id="cb24-7"><a href="#cb24-7" aria-hidden="true" tabindex="-1"></a>       <span class="at">y =</span> <span class="st">"Count"</span>) <span class="sc">+</span></span>
<span id="cb24-8"><a href="#cb24-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Seurat_Visium_tutorial_files/figure-html/unnamed-chunk-13-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="compare-to-negative-binomial-distribution-with-same-mean-and-dispersion" class="level4">
<h4 class="anchored" data-anchor-id="compare-to-negative-binomial-distribution-with-same-mean-and-dispersion">Compare to Negative Binomial Distribution with same mean and dispersion</h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>mu_count <span class="ot">&lt;-</span> <span class="fu">mean</span>(nCount_Spatialdf<span class="sc">$</span>nCount_Spatial)</span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>variance_count <span class="ot">&lt;-</span> <span class="fu">var</span>(nCount_Spatialdf<span class="sc">$</span>nCount_Spatial)</span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (variance_count <span class="sc">&gt;</span> mu_count) {</span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Estimate the dispersion parameter (size)</span></span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a>  size_count <span class="ot">&lt;-</span> mu_count<span class="sc">^</span><span class="dv">2</span> <span class="sc">/</span> (variance_count <span class="sc">-</span> mu_count)</span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(<span class="fu">paste</span>(<span class="st">"Estimated dispersion parameter (size):"</span>, size_count))</span>
<span id="cb25-8"><a href="#cb25-8" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb25-9"><a href="#cb25-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">stop</span>(<span class="st">"Variance must be greater than the mean to estimate size."</span>)</span>
<span id="cb25-10"><a href="#cb25-10" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Estimated dispersion parameter (size): 5.27982237238857"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>nCount_Spatial_rNB<span class="ot">&lt;-</span><span class="fu">rnbinom</span>(<span class="at">n =</span> <span class="fu">length</span>(nCount_Spatialdf<span class="sc">$</span>nCount_Spatial), <span class="at">size=</span>size_count, <span class="at">mu=</span>mu_count)</span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a>nCount_df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">nCount_Spatial1 =</span> nCount_Spatial, <span class="at">nCount_SpatialPoisson =</span> nCount_Spatial_rPoisson, <span class="at">nCountSpatialNB=</span>nCount_Spatial_rNB)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">data =</span> nCount_df, <span class="fu">aes</span>(<span class="at">x =</span> nCount_Spatial, <span class="at">fill =</span> <span class="st">"nCount_Spatial"</span>), <span class="at">binwidth =</span> <span class="dv">1000</span>, <span class="at">alpha =</span> <span class="fl">0.5</span>, <span class="at">position =</span> <span class="st">'identity'</span>) <span class="sc">+</span></span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">data =</span> nCount_df, <span class="fu">aes</span>(<span class="at">x =</span> nCountSpatialNB, <span class="at">fill =</span> <span class="st">"nCountSpatialNB"</span>), <span class="at">binwidth =</span> <span class="dv">1000</span>, <span class="at">alpha =</span> <span class="fl">0.5</span>, <span class="at">position =</span> <span class="st">'identity'</span>) <span class="sc">+</span></span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_manual</span>(<span class="at">name =</span> <span class="st">"Legend"</span>, <span class="at">values =</span> <span class="fu">c</span>(<span class="st">"nCount_Spatial"</span> <span class="ot">=</span> <span class="st">"blue"</span>, <span class="st">"nCountSpatialNB"</span> <span class="ot">=</span> <span class="st">"red"</span>)) <span class="sc">+</span></span>
<span id="cb28-5"><a href="#cb28-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Overlayed Histogram of nCount_Spatial and nCountSpatialNB"</span>, <span class="at">x =</span> <span class="st">"Value"</span>, <span class="at">y =</span> <span class="st">"Count"</span>) <span class="sc">+</span> <span class="fu">theme_minimal</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Seurat_Visium_tutorial_files/figure-html/unnamed-chunk-15-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Checking the variance of our count data vs.&nbsp;the fitted NB model</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sd</span>(nCount_df<span class="sc">$</span>nCount_Spatial1)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 10563.2</code></pre>
</div>
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sd</span>(nCount_df<span class="sc">$</span>nCountSpatialNB)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 10789.6</code></pre>
</div>
</div>
<p>The fitted NB only slightly underestimates the variance (sd=var^2) in the count data! Our count data has a lot of variance, which is why we will perform a normalization with SCTransform to try to remove some of the excess variance.</p>
<hr>
</section>
</section>
<section id="visualize-mitochondrial-contamination-and-filter-it-out" class="level3">
<h3 class="anchored" data-anchor-id="visualize-mitochondrial-contamination-and-filter-it-out">Visualize mitochondrial contamination and filter it out</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>brain[[<span class="st">"percent.mt"</span>]] <span class="ot">&lt;-</span><span class="fu">PercentageFeatureSet</span>(<span class="at">object =</span> brain, <span class="at">pattern =</span> <span class="st">"^MT-|^mt-"</span>)</span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(brain[[<span class="st">"percent.mt"</span>]])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   percent.mt    
 Min.   : 3.652  
 1st Qu.:11.309  
 Median :13.543  
 Mean   :14.016  
 3rd Qu.:15.883  
 Max.   :35.102  </code></pre>
</div>
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>plot_mt <span class="ot">&lt;-</span> <span class="fu">VlnPlot</span>(brain, <span class="at">features =</span> <span class="st">"percent.mt"</span>, <span class="at">pt.size =</span> <span class="fl">0.1</span>) <span class="sc">+</span> <span class="fu">NoLegend</span>()</span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>plot_mt</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Seurat_Visium_tutorial_files/figure-html/unnamed-chunk-17-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>On a histogram, we can see this more clearly:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>brain<span class="sc">@</span>meta.data <span class="sc">%&gt;%</span> <span class="fu">ggplot</span>(<span class="at">mapping=</span><span class="fu">aes</span>(<span class="at">x=</span>percent.mt)) <span class="sc">+</span> </span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">bins=</span><span class="dv">50</span>) <span class="sc">+</span> </span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Seurat_Visium_tutorial_files/figure-html/unnamed-chunk-18-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>brain<span class="sc">@</span>meta.data <span class="sc">%&gt;%</span> <span class="fu">ggplot</span>(<span class="at">mapping=</span><span class="fu">aes</span>(<span class="at">x=</span>percent.mt)) <span class="sc">+</span> </span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">bins=</span><span class="dv">50</span>, <span class="fu">aes</span>(<span class="at">y =</span> <span class="fu">after_stat</span>(density)), <span class="at">colour =</span> <span class="st">"black"</span>,<span class="at">fill =</span> <span class="st">"grey"</span>) <span class="sc">+</span> </span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_density</span>(<span class="at">alpha =</span> <span class="fl">0.5</span>, <span class="at">adjust =</span> <span class="fl">1.0</span>, <span class="at">fill =</span> <span class="st">"#A0CBE8"</span>, <span class="at">colour =</span> <span class="st">"#4E79A7"</span>) <span class="sc">+</span> </span>
<span id="cb37-4"><a href="#cb37-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Seurat_Visium_tutorial_files/figure-html/unnamed-chunk-18-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Here’s our Count Data after filtering out Mitochondrial Genes using Seurat’s <strong>subset()</strong> function</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>?subset</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Help on topic 'subset' was found in the following packages:

  Package               Library
  S4Vectors             /Users/homc/Library/R/arm64/4.4/library
  base                  /Library/Frameworks/R.framework/Resources/library
  data.table            /Users/homc/Library/R/arm64/4.4/library
  BiocGenerics          /Users/homc/Library/R/arm64/4.4/library
  SeuratObject          /Users/homc/Library/R/arm64/4.4/library


Using the first match ...</code></pre>
</div>
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a>brain_mtfilt<span class="ot">&lt;-</span><span class="fu">subset</span>(<span class="at">x =</span> brain, <span class="at">subset =</span> percent.mt <span class="sc">&lt;</span> <span class="dv">28</span>)</span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a><span class="co">#plot1 &lt;- VlnPlot(brain, features = "nCount_Spatial", pt.size = 0.1) + NoLegend()</span></span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a><span class="co">#plot2 &lt;- SpatialFeaturePlot(brain, features = "nCount_Spatial") + theme(legend.position = "right")</span></span>
<span id="cb40-4"><a href="#cb40-4" aria-hidden="true" tabindex="-1"></a><span class="co">#wrap_plots(plot1, plot2)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<hr>
</section>
<section id="normalize-count-data-with-sctransform" class="level3">
<h3 class="anchored" data-anchor-id="normalize-count-data-with-sctransform">Normalize Count Data with scTransform</h3>
<p>Here, we are plotting the library size (molecular counts) across the data and the tissue. We can see there are tissue regions where counts are reduced, coinciding with areas of cortical white matter where neurons are scarce. Where one might normally use LogNormalize() function in Seurat to force each data point to have a standard/median size after normalization, this can introduce problems.</p>
<p>Instead, we will use <strong>scTransform</strong>, which uses regularized negative binomial model for gene expression to account for technical variation while preserving biological variation. ScTransform functions by normalizing data and detecting high variance features and stores this data in the SCT assay.</p>
<p>We can use it with this data, specifying “Spatial” as the assay to perform the transformation on:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="co">#BiocManager::install('glmGamPoi')</span></span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a><span class="do">### fits Gamma-Poisson (aka Negative Binomial) Generalized Linear Models </span></span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a><span class="do">### substantially improves speed for the SCTransform algorithm</span></span>
<span id="cb41-4"><a href="#cb41-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(glmGamPoi)</span>
<span id="cb41-5"><a href="#cb41-5" aria-hidden="true" tabindex="-1"></a>brain_mtfilt <span class="ot">&lt;-</span> <span class="fu">SCTransform</span>(brain_mtfilt, <span class="at">assay =</span> <span class="st">"Spatial"</span>, <span class="at">verbose =</span> <span class="cn">TRUE</span>)</span>
<span id="cb41-6"><a href="#cb41-6" aria-hidden="true" tabindex="-1"></a>brain <span class="ot">&lt;-</span> <span class="fu">SCTransform</span>(brain, <span class="at">assay =</span> <span class="st">"Spatial"</span>, <span class="at">vars.to.regress =</span> <span class="st">"percent.mt"</span>, <span class="at">verbose =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a>brain_mtfilt</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>An object of class Seurat 
48708 features across 2653 samples within 2 assays 
Active assay: SCT (17655 features, 3000 variable features)
 3 layers present: counts, data, scale.data
 1 other assay present: Spatial
 1 spatial field of view present: anterior1</code></pre>
</div>
</div>
<hr>
</section>
<section id="spatial-feature-plot-can-be-used-to-show-the-spatial-expression-pattern-of-a-given-featuregene" class="level3">
<h3 class="anchored" data-anchor-id="spatial-feature-plot-can-be-used-to-show-the-spatial-expression-pattern-of-a-given-featuregene">Spatial Feature Plot can be used to show the spatial expression pattern of a given feature/gene</h3>
<p>The <a href="https://satijalab.org/seurat/reference/spatialplot"><code>SpatialFeaturePlot</code></a> function in Seurat extends <code>FeaturePlot</code>, and can overlay molecular data on top of tissue histology.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="fu">SpatialFeaturePlot</span>(brain_mtfilt, <span class="at">features =</span> <span class="fu">c</span>(<span class="st">"Hpca"</span>, <span class="st">"Ttr"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Seurat_Visium_tutorial_files/figure-html/unnamed-chunk-21-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Seurat’s default parameters emphasize the molecular visualization of the data, but we can adjust the plotting parameters to allow for better visualization of histology. The following parameters can be adjusted:</p>
<ul>
<li><p><code>pt.size.factor</code>- This will scale the size of the spots. Default is 1.6</p></li>
<li><p><code>alpha</code> - minimum and maximum transparency. Default is c(1, 1).</p>
<ul>
<li>try setting to <code>alpha</code> c(0.1, 1), to downweight the transparency of points with lower expression</li>
</ul></li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">SpatialFeaturePlot</span>(brain_mtfilt, <span class="at">features =</span> <span class="st">"Ttr"</span>, <span class="at">pt.size.factor =</span> <span class="dv">2</span>)</span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">SpatialFeaturePlot</span>(brain_mtfilt, <span class="at">features =</span> <span class="st">"Ttr"</span>, <span class="at">alpha =</span> <span class="fu">c</span>(<span class="fl">0.1</span>, <span class="dv">1</span>))</span>
<span id="cb45-3"><a href="#cb45-3" aria-hidden="true" tabindex="-1"></a>p1 <span class="sc">+</span> p2</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Seurat_Visium_tutorial_files/figure-html/unnamed-chunk-22-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ImageFeaturePlot</span>(brain_mtfilt, <span class="at">features=</span><span class="st">"Ttr"</span>, <span class="at">size=</span><span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Seurat_Visium_tutorial_files/figure-html/unnamed-chunk-23-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<hr>
</section>
<section id="dimension-reduction-with-pca-clustering-and-visualize-clusters-with-umap" class="level3">
<h3 class="anchored" data-anchor-id="dimension-reduction-with-pca-clustering-and-visualize-clusters-with-umap">Dimension Reduction with PCA, Clustering and Visualize Clusters with UMAP</h3>
<p>We can then perform dimension reduction with PCA, run k-NN clustering and further visualize with UMAP</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a>brain_mtfilt <span class="ot">&lt;-</span> <span class="fu">RunPCA</span>(brain_mtfilt, <span class="at">assay =</span> <span class="st">"SCT"</span>, <span class="at">verbose =</span> <span class="cn">TRUE</span>)</span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a>brain_mtfilt <span class="ot">&lt;-</span> <span class="fu">FindNeighbors</span>(brain_mtfilt, <span class="at">reduction =</span> <span class="st">"pca"</span>, <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">30</span>)</span>
<span id="cb47-3"><a href="#cb47-3" aria-hidden="true" tabindex="-1"></a>brain_mtfilt <span class="ot">&lt;-</span> <span class="fu">FindClusters</span>(brain_mtfilt, <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span>
<span id="cb47-4"><a href="#cb47-4" aria-hidden="true" tabindex="-1"></a>brain_mtfilt <span class="ot">&lt;-</span> <span class="fu">RunUMAP</span>(brain_mtfilt, <span class="at">reduction =</span> <span class="st">"pca"</span>, <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">30</span>)</span>
<span id="cb47-5"><a href="#cb47-5" aria-hidden="true" tabindex="-1"></a><span class="fu">Reductions</span>(brain_mtfilt)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "pca"  "umap"</code></pre>
</div>
</div>
<p>The PCA analysis resulted in 5 principal components and 14 clusters</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(brain_mtfilt<span class="sc">$</span>seurat_clusters)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  0   1   2   3   4   5   6   7   8   9  10  11  12  13  14 
428 299 257 231 221 208 194 154 154 143 132 130  48  30  24 </code></pre>
</div>
</div>
<p>We can plot the PCA graph along PC1 and PC2 and also show the location of the identified clusters. Note: The use of “label” labels each cluster.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">DimPlot</span>(brain_mtfilt, <span class="at">reduction =</span> <span class="st">"pca"</span>, <span class="at">label =</span> <span class="cn">TRUE</span>)</span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">SpatialDimPlot</span>(brain_mtfilt, <span class="at">label =</span> <span class="cn">TRUE</span>, <span class="at">label.size =</span> <span class="dv">3</span>, <span class="at">pt.size.factor =</span> <span class="fl">2.5</span>)</span>
<span id="cb51-3"><a href="#cb51-3" aria-hidden="true" tabindex="-1"></a>p1 <span class="sc">+</span> p2</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Seurat_Visium_tutorial_files/figure-html/plotPCA-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Then we can visualize clusters via UMAP embedding to show the relative distance between clusters on a 2-dimensional embedding</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">DimPlot</span>(brain_mtfilt, <span class="at">reduction =</span> <span class="st">"umap"</span>, <span class="at">label =</span> <span class="cn">TRUE</span>)</span>
<span id="cb52-2"><a href="#cb52-2" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">SpatialDimPlot</span>(brain_mtfilt, <span class="at">label =</span> <span class="cn">TRUE</span>, <span class="at">label.size =</span> <span class="dv">3</span>, <span class="at">pt.size.factor =</span> <span class="dv">2</span>)</span>
<span id="cb52-3"><a href="#cb52-3" aria-hidden="true" tabindex="-1"></a>p1 <span class="sc">+</span> p2</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Seurat_Visium_tutorial_files/figure-html/unnamed-chunk-26-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<hr>
</section>
<section id="save-point-1" class="level3">
<h3 class="anchored" data-anchor-id="save-point-1">Save Point 1</h3>
<p>It’s always useful to save your Seurat objects as RDS objects for loading again later</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(brain_mtfilt, <span class="at">file =</span> <span class="st">"brain_mtfilt_sctransform.rds"</span>)</span>
<span id="cb53-2"><a href="#cb53-2" aria-hidden="true" tabindex="-1"></a><span class="co">#brain_mtfilt &lt;- readRDS("brain_mtfilt_sctransform.rds")</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<hr>
</section>
<section id="use-spatialdimplot-and-imagedimplot-to-examine-clusters" class="level3">
<h3 class="anchored" data-anchor-id="use-spatialdimplot-and-imagedimplot-to-examine-clusters">Use SpatialDimPlot and ImageDimPlot to examine clusters</h3>
<p><strong>SpatialDimPlot</strong> will show the cluster labels and you can also use ggplot to easily add the X-Y coordinates to the image. You could use <strong>Crop()</strong> to manually crop the coordinates (we won’t here).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a><span class="fu">SpatialDimPlot</span>(brain_mtfilt, <span class="at">label =</span> <span class="cn">TRUE</span>, <span class="at">label.size =</span> <span class="dv">3</span>, <span class="at">pt.size.factor =</span> <span class="dv">2</span>) <span class="sc">+</span>   <span class="fu">theme_minimal</span>() <span class="sc">+</span> </span>
<span id="cb54-2"><a href="#cb54-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">panel.grid.major =</span> <span class="fu">element_line</span>(<span class="at">color =</span> <span class="st">"grey80"</span>, <span class="at">linewidth =</span> <span class="fl">0.5</span>), <span class="co"># Major grid lines</span></span>
<span id="cb54-3"><a href="#cb54-3" aria-hidden="true" tabindex="-1"></a><span class="at">panel.grid.minor =</span> <span class="fu">element_line</span>(<span class="at">color =</span> <span class="st">"grey90"</span>, <span class="at">linewidth =</span> <span class="fl">0.25</span>)) <span class="co"># Minor grid lines</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Seurat_Visium_tutorial_files/figure-html/unnamed-chunk-28-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p><strong>ImageDimPlot</strong> can also be similarly used, but on a spatial map (not overlaid over the image):</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ImageDimPlot</span> (brain_mtfilt, <span class="at">size=</span><span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Seurat_Visium_tutorial_files/figure-html/unnamed-chunk-29-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>SpatialDimPlot and SpatialFeaturePlot both have an interactive option</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a><span class="co">#SpatialDimPlot(brain_mtfilt, label = TRUE, label.size = 3, pt.size.factor = 3, interactive=TRUE)</span></span>
<span id="cb56-2"><a href="#cb56-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-3"><a href="#cb56-3" aria-hidden="true" tabindex="-1"></a><span class="co">#SpatialFeaturePlot(brain_mtfilt, features = "Ttr", alpha = c(0.1, 1), interactive=TRUE)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Another useful interactive feature, <em>LinkedDimPlot</em> links the UMAP representation to the tissue image representation! If you select a cluster in the UMAP plot the corresponding spots in the image will be highlighted.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a><span class="co">#LinkedDimPlot(brain_mtfilt)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The function <em>cells.highlight</em> within SpatialDimPlot can be used to mark a particular cell of interest. This will plot each cluster’s spatial distribution one by one. Here we have clusters 1,2, 3, 4, 5 and 8</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb58"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a><span class="fu">SpatialDimPlot</span>(brain_mtfilt, <span class="at">cells.highlight =</span> <span class="fu">CellsByIdentities</span>(<span class="at">object =</span> brain_mtfilt, <span class="at">idents =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>, <span class="dv">4</span>, <span class="dv">5</span>, <span class="dv">8</span>)), <span class="at">facet.highlight =</span> <span class="cn">TRUE</span>, <span class="at">ncol =</span> <span class="dv">3</span>, <span class="at">pt.size.factor =</span> <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Seurat_Visium_tutorial_files/figure-html/unnamed-chunk-32-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Here are the SCTransformed Counts (UMI) for each Cluster on a Violin Plot</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a>plot1 <span class="ot">&lt;-</span> <span class="fu">VlnPlot</span>(brain_mtfilt, <span class="at">features =</span> <span class="st">"nCount_SCT"</span>, <span class="at">pt.size =</span> <span class="fl">0.1</span>) <span class="sc">+</span> <span class="fu">NoLegend</span>()</span>
<span id="cb59-2"><a href="#cb59-2" aria-hidden="true" tabindex="-1"></a>plot1</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Seurat_Visium_tutorial_files/figure-html/unnamed-chunk-33-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<hr>
</section>
</section>
<section id="find-de-marker-genes-for-clusters" class="level2">
<h2 class="anchored" data-anchor-id="find-de-marker-genes-for-clusters">Find DE Marker Genes for Clusters</h2>
<p>We can identify differentially expressed genes between two groups of cells using a Wilcoxon Rank Sum test with the <em>FindMarkers</em> function. Because the clusters that were found have good spatial restriction, this is one way of finding genes marking spatially distinct regions.</p>
<p>Here we look at DE genes between cluster 2 and cluster 5</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a><span class="co">#devtools::install_github('immunogenomics/presto')</span></span>
<span id="cb60-2"><a href="#cb60-2" aria-hidden="true" tabindex="-1"></a><span class="do">### presto will allow a much faster implementation of the Wilcoxon Rank Sum Test</span></span>
<span id="cb60-3"><a href="#cb60-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(presto)</span>
<span id="cb60-4"><a href="#cb60-4" aria-hidden="true" tabindex="-1"></a>de_markers_2_5 <span class="ot">&lt;-</span> <span class="fu">FindMarkers</span>(brain_mtfilt, <span class="at">ident.1 =</span> <span class="dv">2</span>, <span class="at">ident.2 =</span> <span class="dv">5</span>)</span>
<span id="cb60-5"><a href="#cb60-5" aria-hidden="true" tabindex="-1"></a><span class="fu">SpatialFeaturePlot</span>(<span class="at">object =</span> brain_mtfilt, <span class="at">features =</span> <span class="fu">rownames</span>(de_markers_2_5)[<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>], <span class="at">alpha =</span> <span class="fu">c</span>(<span class="fl">0.1</span>, <span class="dv">1</span>), <span class="at">ncol =</span> <span class="dv">3</span>, <span class="at">pt.size.factor =</span> <span class="fl">2.5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Seurat_Visium_tutorial_files/figure-html/unnamed-chunk-34-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb61"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a>de_markers_2_5_p0<span class="fl">.01</span> <span class="ot">&lt;-</span> de_markers_2_5 <span class="sc">%&gt;%</span> <span class="fu">filter</span>(p_val <span class="sc">&lt;</span> <span class="fl">0.01</span>)</span>
<span id="cb61-2"><a href="#cb61-2" aria-hidden="true" tabindex="-1"></a><span class="fu">SpatialFeaturePlot</span>(<span class="at">object =</span> brain_mtfilt, <span class="at">features =</span> <span class="fu">rownames</span>(de_markers_2_5_p0<span class="fl">.01</span>)[<span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>], <span class="at">alpha =</span> <span class="fu">c</span>(<span class="fl">0.1</span>, <span class="dv">1</span>), <span class="at">ncol =</span> <span class="dv">3</span>, <span class="at">pt.size.factor =</span> <span class="fl">2.5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Seurat_Visium_tutorial_files/figure-html/unnamed-chunk-35-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<hr>
<section id="spatially-variable-genesfeatures-using-morans-i" class="level3">
<h3 class="anchored" data-anchor-id="spatially-variable-genesfeatures-using-morans-i">Spatially Variable Genes/Features using Moran’s I</h3>
<p>A more direct way to find spatially variable genes is using <code>FindSpatiallyVariableFeatures()</code> . It identifies features exhibiting spatial patterning in the absence of pre-annotation. The default method (method=<code>markvariogram</code>) is inspired by the Trendsceek package, which takes STx data as a mark point process and computes a variogram to ID genes based on expression in spatial location. But we can also specify it to use Moran’s I:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb62"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a><span class="co">#install.packages("Rfast2")</span></span>
<span id="cb62-2"><a href="#cb62-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Rfast2)</span>
<span id="cb62-3"><a href="#cb62-3" aria-hidden="true" tabindex="-1"></a>brain_mtfilt <span class="ot">&lt;-</span> <span class="fu">FindSpatiallyVariableFeatures</span>(brain_mtfilt, <span class="at">assay =</span> <span class="st">"SCT"</span>, <span class="at">features =</span> <span class="fu">VariableFeatures</span>(brain_mtfilt)[<span class="dv">1</span><span class="sc">:</span><span class="dv">1000</span>], <span class="at">selection.method =</span> <span class="st">"moransi"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a><span class="co">#rownames(brain_mtfilt$SCT)</span></span>
<span id="cb63-2"><a href="#cb63-2" aria-hidden="true" tabindex="-1"></a><span class="co">#brain_mtfilt$SCT</span></span>
<span id="cb63-3"><a href="#cb63-3" aria-hidden="true" tabindex="-1"></a><span class="co">#names(brain_mtfilt)</span></span>
<span id="cb63-4"><a href="#cb63-4" aria-hidden="true" tabindex="-1"></a><span class="co">#head(brain_mtfilt$SCT@meta.features)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb64"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a>moransi_output_df <span class="ot">&lt;-</span> brain_mtfilt<span class="sc">@</span>assays<span class="sc">$</span>SCT<span class="sc">@</span>meta.features <span class="sc">%&gt;%</span></span>
<span id="cb64-2"><a href="#cb64-2" aria-hidden="true" tabindex="-1"></a>  na.exclude</span>
<span id="cb64-3"><a href="#cb64-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(moransi_output_df[<span class="fu">order</span>(moransi_output_df<span class="sc">$</span>MoransI_observed, <span class="at">decreasing =</span> T), ])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>       MoransI_observed MoransI_p.value moransi.spatially.variable
Calb2         0.6785197    0.0009756098                       TRUE
Gng4          0.6585290    0.0009756098                       TRUE
Ttr           0.6213368    0.0009756098                       TRUE
Nrgn          0.6171936    0.0009756098                       TRUE
S100a5        0.6142008    0.0009756098                       TRUE
Doc2g         0.6089830    0.0009756098                       TRUE
       moransi.spatially.variable.rank
Calb2                                1
Gng4                                 2
Ttr                                  3
Nrgn                                 4
S100a5                               5
Doc2g                                6</code></pre>
</div>
</div>
<p>See for reference (https://github.com/satijalab/seurat/issues/7729)</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb66"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a>top.clusters <span class="ot">&lt;-</span> <span class="fu">rownames</span>(brain_mtfilt<span class="sc">$</span>SCT<span class="sc">@</span>meta.features)[<span class="fu">which</span>(brain_mtfilt<span class="sc">$</span>SCT<span class="sc">@</span>meta.features<span class="sc">$</span>moransi.spatially.variable.rank <span class="sc">&lt;</span> <span class="dv">6</span>)]</span>
<span id="cb66-2"><a href="#cb66-2" aria-hidden="true" tabindex="-1"></a><span class="fu">SpatialFeaturePlot</span>(brain_mtfilt, <span class="at">features =</span> top.clusters, <span class="at">ncol =</span> <span class="dv">3</span>, <span class="at">alpha =</span> <span class="fu">c</span>(<span class="fl">0.1</span>, <span class="dv">1</span>), <span class="at">pt.size.factor =</span> <span class="fl">2.5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Seurat_Visium_tutorial_files/figure-html/unnamed-chunk-39-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<hr>
</section>
</section>
<section id="subset-cortex-anatomical-region" class="level2">
<h2 class="anchored" data-anchor-id="subset-cortex-anatomical-region">Subset cortex anatomical region</h2>
<p>Here we can do so by selecting specific clusters that correspond to this anatomical region</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb67"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a>cortex <span class="ot">&lt;-</span> <span class="fu">subset</span>(brain_mtfilt, <span class="at">idents =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>, <span class="dv">4</span>, <span class="dv">6</span>))</span>
<span id="cb67-2"><a href="#cb67-2" aria-hidden="true" tabindex="-1"></a><span class="fu">SpatialDimPlot</span>(cortex, <span class="at">crop =</span> <span class="cn">TRUE</span>, <span class="at">label =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Seurat_Visium_tutorial_files/figure-html/subset_cortex-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>After subsetting we should renormalize cortex spatial data</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb68"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb68-1"><a href="#cb68-1" aria-hidden="true" tabindex="-1"></a>cortex <span class="ot">&lt;-</span> <span class="fu">SCTransform</span>(cortex, <span class="at">assay =</span> <span class="st">"Spatial"</span>, <span class="at">verbose =</span> <span class="cn">FALSE</span>) <span class="sc">%&gt;%</span></span>
<span id="cb68-2"><a href="#cb68-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">RunPCA</span>(<span class="at">verbose =</span> <span class="cn">FALSE</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb68-3"><a href="#cb68-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">FindNeighbors</span>(<span class="at">reduction =</span> <span class="st">"pca"</span>, <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">30</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb68-4"><a href="#cb68-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">FindClusters</span>(<span class="at">verbose =</span> <span class="cn">FALSE</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb68-5"><a href="#cb68-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">RunUMAP</span>(<span class="at">reduction =</span> <span class="st">"pca"</span>, <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">30</span>)</span>
<span id="cb68-6"><a href="#cb68-6" aria-hidden="true" tabindex="-1"></a><span class="fu">Reductions</span>(cortex)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "pca"  "umap"</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb70"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb70-1"><a href="#cb70-1" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">DimPlot</span>(cortex, <span class="at">reduction =</span> <span class="st">"umap"</span>, <span class="at">label =</span> <span class="cn">TRUE</span>)</span>
<span id="cb70-2"><a href="#cb70-2" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">SpatialDimPlot</span>(cortex, <span class="at">label =</span> <span class="cn">TRUE</span>, <span class="at">label.size =</span> <span class="dv">3</span>, <span class="at">pt.size.factor =</span> <span class="fl">2.5</span>)</span>
<span id="cb70-3"><a href="#cb70-3" aria-hidden="true" tabindex="-1"></a>p1 <span class="sc">+</span> p2</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Seurat_Visium_tutorial_files/figure-html/unnamed-chunk-41-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<hr>
<section id="annotation-using-with-a-scrna-seq-reference" class="level3">
<h3 class="anchored" data-anchor-id="annotation-using-with-a-scrna-seq-reference">Annotation using with a scRNA-seq reference</h3>
<p>We will use an algorithm called Robust Cell Type Decomposition to annotate cell types in the query STx dataset using labels from a reference scRNA-seq dataset</p>
<p>First, install RCTD and load it using <strong>library(spacexr)</strong> (skip if done ahead of time)</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb71"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a><span class="co">#if (!requireNamespace("spacexr", quietly = TRUE)) {</span></span>
<span id="cb71-2"><a href="#cb71-2" aria-hidden="true" tabindex="-1"></a><span class="co">#  devtools::install_github("dmcable/spacexr", build_vignettes = FALSE)</span></span>
<span id="cb71-3"><a href="#cb71-3" aria-hidden="true" tabindex="-1"></a><span class="co">#}</span></span>
<span id="cb71-4"><a href="#cb71-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-5"><a href="#cb71-5" aria-hidden="true" tabindex="-1"></a><span class="do">### Note: I had issues installing until I took myself off NIH VPN. Might want to do the same if you encounter the same issue</span></span>
<span id="cb71-6"><a href="#cb71-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-7"><a href="#cb71-7" aria-hidden="true" tabindex="-1"></a><span class="co">#remotes::install_github("dmcable/RCTD")</span></span>
<span id="cb71-8"><a href="#cb71-8" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(spacexr)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Download mouse cortex reference <a href="https://www.dropbox.com/scl/fi/r1mixf4eof2cot891n215/allen_scRNAseq_ref.Rds?rlkey=ynr6s6wu1efqsjsu3h40vitt7&amp;dl=0">here</a>. (skip if done ahead of time) This reference scRNAseq dataset has been reduced to 200,000 cells (and rare cell types fewer than 25 cells have been removed).</p>
<p>Load the mouse cortex reference as <code>ref</code> object:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb72"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb72-1"><a href="#cb72-1" aria-hidden="true" tabindex="-1"></a><span class="co">#setwd("/Users/homc/Library/CloudStorage/OneDrive-NationalInstitutesofHealth/MHO_WORK/MyWorkshops/SpatialTranscriptomics/Seurat_tutorial/")</span></span>
<span id="cb72-2"><a href="#cb72-2" aria-hidden="true" tabindex="-1"></a>ref <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"allen_cortex.rds"</span>)</span>
<span id="cb72-3"><a href="#cb72-3" aria-hidden="true" tabindex="-1"></a>ref <span class="ot">&lt;-</span> <span class="fu">UpdateSeuratObject</span>(ref)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb73"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb73-1"><a href="#cb73-1" aria-hidden="true" tabindex="-1"></a>ref_counts <span class="ot">&lt;-</span> ref[[<span class="st">"RNA"</span>]]<span class="sc">$</span>counts</span>
<span id="cb73-2"><a href="#cb73-2" aria-hidden="true" tabindex="-1"></a>ref_cluster <span class="ot">&lt;-</span> <span class="fu">as.factor</span>(ref<span class="sc">$</span>subclass)</span>
<span id="cb73-3"><a href="#cb73-3" aria-hidden="true" tabindex="-1"></a>ref_nUMI <span class="ot">&lt;-</span> ref<span class="sc">$</span>nCount_RNA</span>
<span id="cb73-4"><a href="#cb73-4" aria-hidden="true" tabindex="-1"></a><span class="fu">levels</span>(ref_cluster) <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">"/"</span>, <span class="st">"-"</span>, <span class="fu">levels</span>(ref_cluster))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Sketch the cortical subset of the Visium dataset, which downsamples the high-dimensional spatial RNA expression data, which can help with scalability for large datasets, to 10,000 cells</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb74"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb74-1"><a href="#cb74-1" aria-hidden="true" tabindex="-1"></a><span class="fu">DefaultAssay</span>(cortex) <span class="ot">&lt;-</span> <span class="st">"Spatial"</span></span>
<span id="cb74-2"><a href="#cb74-2" aria-hidden="true" tabindex="-1"></a>cortex <span class="ot">&lt;-</span> <span class="fu">FindVariableFeatures</span>(cortex)</span>
<span id="cb74-3"><a href="#cb74-3" aria-hidden="true" tabindex="-1"></a>cortex <span class="ot">&lt;-</span> <span class="fu">NormalizeData</span>(cortex)</span>
<span id="cb74-4"><a href="#cb74-4" aria-hidden="true" tabindex="-1"></a>cortex <span class="ot">&lt;-</span> <span class="fu">ScaleData</span>(cortex)</span>
<span id="cb74-5"><a href="#cb74-5" aria-hidden="true" tabindex="-1"></a>cortex <span class="ot">&lt;-</span> <span class="fu">SketchData</span>(</span>
<span id="cb74-6"><a href="#cb74-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">object =</span> cortex,</span>
<span id="cb74-7"><a href="#cb74-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">ncells =</span> <span class="dv">10000</span>,</span>
<span id="cb74-8"><a href="#cb74-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">method =</span> <span class="st">"LeverageScore"</span>,</span>
<span id="cb74-9"><a href="#cb74-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">sketched.assay =</span> <span class="st">"sketch"</span></span>
<span id="cb74-10"><a href="#cb74-10" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb75"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb75-1"><a href="#cb75-1" aria-hidden="true" tabindex="-1"></a><span class="fu">DefaultAssay</span>(cortex) <span class="ot">&lt;-</span> <span class="st">"sketch"</span></span>
<span id="cb75-2"><a href="#cb75-2" aria-hidden="true" tabindex="-1"></a>cortex <span class="ot">&lt;-</span> <span class="fu">FindVariableFeatures</span>(cortex, <span class="at">assay=</span><span class="st">"sketch"</span>)</span>
<span id="cb75-3"><a href="#cb75-3" aria-hidden="true" tabindex="-1"></a>cortex <span class="ot">&lt;-</span> <span class="fu">ScaleData</span>(cortex, <span class="at">assay=</span><span class="st">"sketch"</span>)</span>
<span id="cb75-4"><a href="#cb75-4" aria-hidden="true" tabindex="-1"></a>cortex <span class="ot">&lt;-</span> <span class="fu">RunPCA</span>(cortex, <span class="at">assay =</span> <span class="st">"sketch"</span>, <span class="at">reduction.name =</span> <span class="st">"pca.cortex.sketch"</span>, <span class="at">verbose =</span> T)</span>
<span id="cb75-5"><a href="#cb75-5" aria-hidden="true" tabindex="-1"></a>cortex <span class="ot">&lt;-</span> <span class="fu">FindNeighbors</span>(cortex, <span class="at">reduction =</span> <span class="st">"pca.cortex.sketch"</span>, <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">50</span>)</span>
<span id="cb75-6"><a href="#cb75-6" aria-hidden="true" tabindex="-1"></a>cortex <span class="ot">&lt;-</span> <span class="fu">RunUMAP</span>(cortex, <span class="at">reduction =</span> <span class="st">"pca.cortex.sketch"</span>, <span class="at">reduction.name =</span> <span class="st">"umap.cortex.sketch"</span>, <span class="at">return.model =</span> T, <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">50</span>, <span class="at">verbose =</span> T)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Create RCTD reference object</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb76"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb76-1"><a href="#cb76-1" aria-hidden="true" tabindex="-1"></a><span class="do">### make reference object from the ref scRNA-seq object</span></span>
<span id="cb76-2"><a href="#cb76-2" aria-hidden="true" tabindex="-1"></a>reference <span class="ot">&lt;-</span> <span class="fu">Reference</span>(ref_counts, ref_cluster, ref_nUMI) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Create the RCTD query object</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb77"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb77-1"><a href="#cb77-1" aria-hidden="true" tabindex="-1"></a><span class="do">### make query SpatialRNA object from the cortex Visium data</span></span>
<span id="cb77-2"><a href="#cb77-2" aria-hidden="true" tabindex="-1"></a>query_counts_hd <span class="ot">&lt;-</span> cortex[[<span class="st">"sketch"</span>]]<span class="sc">$</span>counts</span>
<span id="cb77-3"><a href="#cb77-3" aria-hidden="true" tabindex="-1"></a>query_cells_hd <span class="ot">&lt;-</span> <span class="fu">colnames</span>(cortex[[<span class="st">"sketch"</span>]])</span>
<span id="cb77-4"><a href="#cb77-4" aria-hidden="true" tabindex="-1"></a>query_coords <span class="ot">&lt;-</span> <span class="fu">GetTissueCoordinates</span>(cortex)[query_cells_hd, <span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>]</span>
<span id="cb77-5"><a href="#cb77-5" aria-hidden="true" tabindex="-1"></a>query <span class="ot">&lt;-</span> <span class="fu">SpatialRNA</span>(query_coords, query_counts_hd, <span class="fu">colSums</span>(query_counts_hd))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb78"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb78-1"><a href="#cb78-1" aria-hidden="true" tabindex="-1"></a><span class="co"># run RCTD on query and reference</span></span>
<span id="cb78-2"><a href="#cb78-2" aria-hidden="true" tabindex="-1"></a>RCTD <span class="ot">&lt;-</span> <span class="fu">create.RCTD</span>(query, reference, <span class="at">max_cores =</span> <span class="dv">28</span>, <span class="at">CELL_MIN_INSTANCE=</span><span class="dv">5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
     Astro         CR       Endo    L2-3 IT         L4      L5 IT      L5 PT 
       368          7         94        982       1401        880        544 
     L6 CT      L6 IT        L6b      Lamp5 Macrophage      Meis2         NP 
       960       1872        358       1122         51         45        362 
     Oligo       Peri      Pvalb   Serpinf1        SMC       Sncg        Sst 
        91         32       1337         27         55        125       1741 
       Vip       VLMC 
      1728         67 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb80"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb80-1"><a href="#cb80-1" aria-hidden="true" tabindex="-1"></a>RCTD <span class="ot">&lt;-</span> <span class="fu">run.RCTD</span>(RCTD, <span class="at">doublet_mode =</span> <span class="st">"doublet"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "gather_results: finished 1000"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb82"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb82-1"><a href="#cb82-1" aria-hidden="true" tabindex="-1"></a><span class="co"># add results back to cortex Seurat object</span></span>
<span id="cb82-2"><a href="#cb82-2" aria-hidden="true" tabindex="-1"></a>cortex <span class="ot">&lt;-</span> <span class="fu">AddMetaData</span>(cortex, <span class="at">metadata =</span> RCTD<span class="sc">@</span>results<span class="sc">$</span>results_df)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Project RCTD labels from sketched cortical cells to all cortical cells</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb83"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb83-1"><a href="#cb83-1" aria-hidden="true" tabindex="-1"></a>cortex<span class="sc">$</span>first_type <span class="ot">&lt;-</span> <span class="fu">as.character</span>(cortex<span class="sc">$</span>first_type)</span>
<span id="cb83-2"><a href="#cb83-2" aria-hidden="true" tabindex="-1"></a>cortex<span class="sc">$</span>first_type[<span class="fu">is.na</span>(cortex<span class="sc">$</span>first_type)] <span class="ot">&lt;-</span> <span class="st">"Unknown"</span></span>
<span id="cb83-3"><a href="#cb83-3" aria-hidden="true" tabindex="-1"></a>cortex <span class="ot">&lt;-</span> <span class="fu">ProjectData</span>(</span>
<span id="cb83-4"><a href="#cb83-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">object =</span> cortex,</span>
<span id="cb83-5"><a href="#cb83-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">assay =</span> <span class="st">"Spatial"</span>,</span>
<span id="cb83-6"><a href="#cb83-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">full.reduction =</span> <span class="st">"pca"</span>,</span>
<span id="cb83-7"><a href="#cb83-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">sketched.assay =</span> <span class="st">"sketch"</span>,</span>
<span id="cb83-8"><a href="#cb83-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">sketched.reduction =</span> <span class="st">"pca.cortex.sketch"</span>,</span>
<span id="cb83-9"><a href="#cb83-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">umap.model =</span> <span class="st">"umap.cortex.sketch"</span>,</span>
<span id="cb83-10"><a href="#cb83-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">50</span>,</span>
<span id="cb83-11"><a href="#cb83-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">refdata =</span> <span class="fu">list</span>(<span class="at">full_first_type =</span> <span class="st">"first_type"</span>)</span>
<span id="cb83-12"><a href="#cb83-12" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb84"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb84-1"><a href="#cb84-1" aria-hidden="true" tabindex="-1"></a><span class="co">#plot_all_cell_types(RCTD@results$results_df, coords, as.vector(cortex$first_type), resultsdir="/Users/homc/Library/CloudStorage/OneDrive-NationalInstitutesofHealth/MHO_WORK/MyWorkshops/SpatialTranscriptomics/Seurat_Visium_tutorial/out")</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb85"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb85-1"><a href="#cb85-1" aria-hidden="true" tabindex="-1"></a><span class="fu">DefaultAssay</span>(brain_mtfilt) <span class="ot">&lt;-</span> <span class="st">"Spatial"</span></span>
<span id="cb85-2"><a href="#cb85-2" aria-hidden="true" tabindex="-1"></a><span class="co"># we only ran RCTD on the cortical cells</span></span>
<span id="cb85-3"><a href="#cb85-3" aria-hidden="true" tabindex="-1"></a><span class="co"># set labels to all other cells as "Unknown"</span></span>
<span id="cb85-4"><a href="#cb85-4" aria-hidden="true" tabindex="-1"></a>brain_mtfilt[[]][, <span class="st">"full_first_type"</span>] <span class="ot">&lt;-</span> <span class="st">"Unknown"</span></span>
<span id="cb85-5"><a href="#cb85-5" aria-hidden="true" tabindex="-1"></a>brain_mtfilt<span class="sc">$</span>full_first_type[<span class="fu">Cells</span>(cortex)] <span class="ot">&lt;-</span> cortex<span class="sc">$</span>full_first_type[<span class="fu">Cells</span>(cortex)]</span>
<span id="cb85-6"><a href="#cb85-6" aria-hidden="true" tabindex="-1"></a><span class="fu">Idents</span>(brain_mtfilt) <span class="ot">&lt;-</span> <span class="st">"full_first_type"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="save-point-2" class="level3">
<h3 class="anchored" data-anchor-id="save-point-2">Save Point 2</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb86"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb86-1"><a href="#cb86-1" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(brain_mtfilt, <span class="at">file =</span> <span class="st">"brain_mtfilt_sctransform_cortexannot.rds"</span>)</span>
<span id="cb86-2"><a href="#cb86-2" aria-hidden="true" tabindex="-1"></a><span class="co">#brain_mtfilt &lt;- readRDS("brain_mtfilt_sctransform_cortexannot.rds")</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="visualize-excitatory-interneuron-layers" class="level4">
<h4 class="anchored" data-anchor-id="visualize-excitatory-interneuron-layers">Visualize excitatory interneuron layers</h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb87"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb87-1"><a href="#cb87-1" aria-hidden="true" tabindex="-1"></a><span class="co"># now we can spatially map the location of any scRNA-seq cell type</span></span>
<span id="cb87-2"><a href="#cb87-2" aria-hidden="true" tabindex="-1"></a><span class="co"># start with Layered (starts with L), excitatory neurons in the cortex</span></span>
<span id="cb87-3"><a href="#cb87-3" aria-hidden="true" tabindex="-1"></a>cells <span class="ot">&lt;-</span> <span class="fu">CellsByIdentities</span>(brain_mtfilt)</span>
<span id="cb87-4"><a href="#cb87-4" aria-hidden="true" tabindex="-1"></a>excitatory_names <span class="ot">&lt;-</span> <span class="fu">sort</span>(<span class="fu">grep</span>(<span class="st">"^L.*"</span>, <span class="fu">names</span>(cells), <span class="at">value =</span> <span class="cn">TRUE</span>))</span>
<span id="cb87-5"><a href="#cb87-5" aria-hidden="true" tabindex="-1"></a><span class="fu">SpatialDimPlot</span>(brain_mtfilt, <span class="at">cells.highlight =</span> cells[excitatory_names], <span class="at">cols.highlight =</span> <span class="fu">c</span>(<span class="st">"#FFFF00"</span>, <span class="st">"grey50"</span>), <span class="at">facet.highlight =</span> T, <span class="at">combine =</span> T, <span class="at">ncol =</span> <span class="dv">4</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Seurat_Visium_tutorial_files/figure-html/unnamed-chunk-54-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb88"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb88-1"><a href="#cb88-1" aria-hidden="true" tabindex="-1"></a><span class="do">### Here we can visualize all the cells</span></span>
<span id="cb88-2"><a href="#cb88-2" aria-hidden="true" tabindex="-1"></a><span class="co">#all_names&lt;- names(cells)</span></span>
<span id="cb88-3"><a href="#cb88-3" aria-hidden="true" tabindex="-1"></a><span class="co">#SpatialDimPlot(brain_mtfilt, cells.highlight = cells[all_names], cols.highlight = c("#FFFF00", "grey50"), facet.highlight = T, combine = F, ncol = 1)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb89"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb89-1"><a href="#cb89-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(viridis)</span>
<span id="cb89-2"><a href="#cb89-2" aria-hidden="true" tabindex="-1"></a>brain_mtfilt_types <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">table</span>(brain_mtfilt<span class="sc">$</span>full_first_type))</span>
<span id="cb89-3"><a href="#cb89-3" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(brain_mtfilt_types, <span class="fu">aes</span>(<span class="at">x =</span> Var1, <span class="at">y =</span> Freq, <span class="at">fill=</span>Var1)) <span class="sc">+</span></span>
<span id="cb89-4"><a href="#cb89-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="at">stat =</span> <span class="st">"identity"</span>) <span class="sc">+</span></span>
<span id="cb89-5"><a href="#cb89-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>() <span class="sc">+</span></span>
<span id="cb89-6"><a href="#cb89-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Assigned Cell Type after RTCD"</span>, <span class="at">x =</span> <span class="st">"Assigned Cell Type after RTCD"</span>, <span class="at">y =</span> <span class="st">"Count"</span>) <span class="sc">+</span></span>
<span id="cb89-7"><a href="#cb89-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_viridis</span>(<span class="at">discrete=</span>T, <span class="at">option=</span><span class="st">"C"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Seurat_Visium_tutorial_files/figure-html/unnamed-chunk-56-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p><strong>ImageDimPlot</strong> will allow us to see the annotated cell identities on a spatial map:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb90"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb90-1"><a href="#cb90-1" aria-hidden="true" tabindex="-1"></a><span class="fu">DefaultAssay</span>(brain_mtfilt) <span class="ot">&lt;-</span> <span class="st">"Spatial"</span></span>
<span id="cb90-2"><a href="#cb90-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ImageDimPlot</span>(brain_mtfilt, <span class="at">size=</span><span class="dv">2</span>, <span class="at">dark.background=</span><span class="cn">FALSE</span>, <span class="at">cols=</span><span class="st">"polychrome"</span>) <span class="sc">+</span>   <span class="fu">theme_minimal</span>() <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">panel.grid.major =</span> <span class="fu">element_line</span>(<span class="at">color =</span> <span class="st">"grey80"</span>, <span class="at">linewidth =</span> <span class="fl">0.5</span>), <span class="co"># Major grid lines</span></span>
<span id="cb90-3"><a href="#cb90-3" aria-hidden="true" tabindex="-1"></a><span class="at">panel.grid.minor =</span> <span class="fu">element_line</span>(<span class="at">color =</span> <span class="st">"grey90"</span>, <span class="at">linewidth =</span> <span class="fl">0.25</span>)) <span class="sc">+</span> <span class="co"># Minor grid lines + </span></span>
<span id="cb90-4"><a href="#cb90-4" aria-hidden="true" tabindex="-1"></a><span class="fu">scale_x_continuous</span>(<span class="at">name =</span> <span class="st">"X Coordinate"</span>) <span class="sc">+</span> <span class="co"># X-axis label</span></span>
<span id="cb90-5"><a href="#cb90-5" aria-hidden="true" tabindex="-1"></a><span class="fu">scale_y_continuous</span>(<span class="at">name =</span> <span class="st">"Y Coordinate"</span>) <span class="co"># Y-axis label</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Seurat_Visium_tutorial_files/figure-html/unnamed-chunk-57-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
</section>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>